RavenDB-25420 Load & Lazy loads bugfixes

maciejaszyk opened the pull request:
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-25420

### Additional description

Includes:
 - LazyLoad - Ensure all includes are loaded when loading "load" document from session.
 - Load - Consider parameters other than `Includes` when skipping a request to the server.

### Type of change

- [ ] Bug fix
- [ ] Regression bug fix
- [ ] Optimization
- [ ] New feature

### How risky is the change?

- [ ] Low 
- [ ] Moderate 
- [ ] High
- [ ] Not relevant

### Backward compatibility

- [ ] Non breaking change
- [ ] Ensured. Please explain how has it been implemented?
- [ ] Breaking change
- [ ] Not relevant

### Is it platform specific issue?

- [ ] Yes. Please list the affected platforms.
- [ ] No

### Documentation update

- [ ] This change requires a documentation update. Please mark the issue on YouTrack using `Documentation Required` tag.
- [ ] No documentation update is needed 

### Testing by Contributor

- [ ] Tests have been added that prove the fix is effective or that the feature works
- [ ] Internal classes added to the test class (e.g. entity or index definition classes) have the lowest possible access modifier (preferable `private`) 
- [ ] It has been verified by manual testing
- [ ] Existing tests verify the correct behavior

### Testing by RavenDB QA team

- [ ] This change requires a special QA testing due to possible performance or resources usage implications (CPU, memory, IO). Please mark the issue on YouTrack using `QA Required` tag.
- [ ] No special testing by RavenDB QA team is needed

### Is there any existing behavior change of other features due to this change?

- [ ] Yes. Please list the affected features/subsystems and provide appropriate explanation
- [ ] No

### UI work

- [ ] It requires further work in the Studio. Please mark the issue on YouTrack using `Studio Required` tag.
- [ ] No UI work is needed

ayende commented on src/Raven.Client/Documents/Session/Operations/Lazy/LazyLoadOperation.cs:
```diff
@@ -32,28 +31,9 @@ public GetRequest CreateRequest(JsonOperationContext ctx)
-            bool hasItems = false;
             foreach (var id in _ids)
-            {
-                if (_session.IsLoadedOrDeleted(id))
```
No, you are killing the feature for this scenario.
If we already has it in the session, we don't need to go anywhere.
And if _all_ are in the session, is is fine.

The key problem is that we have `include` here.
So we need to also check if it is in the session, if the included session is also in the session.

maciejaszyk committed:
RavenDB-25420
 - LazyLoad - Ensure all includes are loaded when loading "load" document from session.
 - Load - Consider parameters other than `Includes` when skipping a request to the server.

arekpalinski commented on src/Raven.Client/Documents/Session/InMemoryDocumentSessionOperations.cs:
```diff
@@ -2278,6 +2279,60 @@ public bool CheckIfIdAlreadyIncluded(string[] ids, IEnumerable<string> includes)
         }
 
+        public IEnumerable<(string Id, bool ExistsInSession)> GetDocumentsRequiredToLoad<TEnumerable>(string[] ids, TEnumerable includes)
+        where TEnumerable : IEnumerable<string>
```
Please consider to use this method in `CheckIfIdAlreadyIncluded` to avoid duplicated code

maciejaszyk replied:
done

maciejaszyk committed:
RavenDB-25420 Reuse method logic.

maciejaszyk committed:
RavenDB-25420
- Provide a list of counters to include.
- Correct NumberOfRequests in test.

maciejaszyk commented on src/Raven.Client/Documents/Session/InMemoryDocumentSessionOperations.cs:
```diff
@@ -1621,7 +1621,7 @@ internal void RegisterCounters(BlittableJsonReaderObject resultCounters, string[
             else
             {
-                RegisterCountersInternal(resultCounters, countersToInclude: null, fromQueryResult: false, gotAll: gotAll);
+                RegisterCountersInternal(resultCounters, countersToInclude: ids.ToDictionary(x => x, _ => countersToInclude), fromQueryResult: false, gotAll: gotAll);
```
@aviv86 can you take a look at this fix? There was a problem whenever a counter was in the session (and cached). We were getting an NRE on line 1693.

ayende replied:
Do we properly handle scenarios where the user wanted to include a counter / time series range?

maciejaszyk replied:
I've created a ticket: https://issues.hibernatingrhinos.com/issue/RavenDB-25455/Counter-is-overridden-in-Load-operation

ayende commented on src/Raven.Client/Documents/Session/InMemoryDocumentSessionOperations.cs:
```diff
@@ -2248,36 +2249,72 @@ internal bool CheckIfRevisionByDateTimeBeforeAlreadyIncluded(string id, DateTime
+            return true;
+        }
+
+        public IEnumerable<(string Id, bool ExistsInSession)> GetDocumentsRequiredToLoad<TEnumerable>(string[] ids, TEnumerable includes)
```
This is about document ref includes, what about counters, time series ranges, etc?

maciejaszyk replied:
This is not implemented. If a user requires time series, counters, or compare exchange, we send the request anyway. Previously (in normal Load), we did not send any request if all requested documents were in the session (without checking if we had all includes in the session). That led to an increased `NumberOfRequests` later because we hadn't loaded anything.

ayende replied:
I'm fine with saying that include on those would require a separate call, yes.

ayende commented on test/SlowTests/Client/Lazy/Async/RavenDB-25420.cs:
```diff
@@ -0,0 +1,348 @@
+        Assert.Equal(1, session.Advanced.NumberOfRequests);
+
+        var baseLineOffset = _baseLine.AddMinutes(5);
+        await session.LoadAsync<Book>([book1.Id, book2.Id, book3.Id], builder => builder.IncludeTimeSeries(TimeSeriesName, _baseLine, baseLineOffset));
```
Please also check that this works properly when we run on a _different_ range.

ppekrol commented:
Requires https://issues.hibernatingrhinos.com/issue/RavenDB-25455/Counter-is-overridden-in-Load-operation. before merging

arekpalinski commented:
@ravendb-bot test this in debug please

Commit:
RavenDB-25455 : Fixes in counters include cache merging and session tracking :
- Fix NRE when including counters on documents that already have tracked counters (pass include-map to registration + null-safe handling).
- Make session counter cache reflect deferred increments immediately (read-your-writes before SaveChanges).
- Prevent included counter results from overwriting newer in-session counter values by merging server values with deferred counter ops.

Commit:
RavenDB-25455 : fix 2 counter tests - it was wrong to expect the cached-counter value to NOT reflect pending changes made inside the session. incs/deletes in tracked counters should be reflected in cache, even before SaveChanges - same as in tracked entities.
also, deleting a counter should set it's cached value to null (known missing) so that we wont go to server again to fetch it.

Commit:
RavenDB-25455 : address PR comments
